compiler/qsc_data_structures/src/line_column/tests.rs

```diff
@@ -0,0 +1,398 @@
+    // code points  | e7       e7       e7
+    // utf-8 units  | c3a7     c3a7     c3a7
+    // utf-16 units | e7       e7       e7
+    let contents = "ççç\nççç";
+    check_all_offsets(
+        contents,
+        &expect![[r"
+            byte | utf-8 | utf-16 | char
+               0 |  0, 0 |  0, 0  | 'ç'
+               1 |  0, 2 |  0, 1  |
+               2 |  0, 2 |  0, 1  | 'ç'
+               3 |  0, 4 |  0, 2  |
+               4 |  0, 4 |  0, 2  | 'ç'
+               5 |  0, 6 |  0, 3  |
+               6 |  0, 6 |  0, 3  | '\n'
+               7 |  1, 0 |  1, 0  | 'ç'
+               8 |  1, 2 |  1, 1  |
+               9 |  1, 2 |  1, 1  | 'ç'
+              10 |  1, 4 |  1, 2  |
+              11 |  1, 4 |  1, 2  | 'ç'
+              12 |  1, 6 |  1, 3  |
+              13 |  1, 6 |  1, 3  | <eof>
+        "]],
+    );
+}
+
+#[test]
+fn utf_8_multibyte_utf_16_surrogate() {
+    // both encodings have multi-unit characters
+    // string       | 𝑓𝑓
+    // chars        | 𝑓                 𝑓
+    // code points  | 1d453             1d453
+    // utf-8 units  | f09d9193          f09d9193
+    // utf-16 units | d835     dc53     d835     dc53
+
+    let contents = "𝑓𝑓\n𝑓𝑓";
+    check_all_offsets(
+        contents,
+        &expect![[r"
+            byte | utf-8 | utf-16 | char
+               0 |  0, 0 |  0, 0  | '𝑓'
+               1 |  0, 4 |  0, 2  |
+               2 |  0, 4 |  0, 2  |
+               3 |  0, 4 |  0, 2  |
+               4 |  0, 4 |  0, 2  | '𝑓'
+               5 |  0, 8 |  0, 4  |
+               6 |  0, 8 |  0, 4  |
+               7 |  0, 8 |  0, 4  |
+               8 |  0, 8 |  0, 4  | '\n'
+               9 |  1, 0 |  1, 0  | '𝑓'
+              10 |  1, 4 |  1, 2  |
+              11 |  1, 4 |  1, 2  |
+              12 |  1, 4 |  1, 2  |
+              13 |  1, 4 |  1, 2  | '𝑓'
+              14 |  1, 8 |  1, 4  |
+              15 |  1, 8 |  1, 4  |
+              16 |  1, 8 |  1, 4  |
+              17 |  1, 8 |  1, 4  | <eof>
+        "]],
+    );
+}
+
+#[test]
+fn grapheme_clusters() {
+    // grapheme clusters, both encodings have multi-unit characters
+    // string       | 𝑓(𝑥⃗) ≔ Σᵢ 𝑥ᵢ 𝑟ᵢ
+    // chars        | 𝑓                 (        𝑥                 ⃗        )                 ≔                 Σ        ᵢ                 𝑥                 ᵢ                 𝑟                 ᵢ
+    // code points  | 1d453             28       1d465             20d7     29       20       2254     20       3a3      1d62     20       1d465             1d62     20       1d45f             1d62
+    // utf-8 units  | f09d9193          28       f09d91a5          e28397   29       20       e28994   20       cea3     e1b5a2   20       f09d91a5          e1b5a2   20       f09d919f          e1b5a2
+    // utf-16 units | d835     dc53     28       d835     dc65     20d7     29       20       2254     20       3a3      1d62     20       d835     dc65     1d62     20       d835     dc5f     1d62
```

Comment:
Again, if these are units, should probably make each be 2 bytes long (0028 instead of 28) so that it reads the same as the byte offset/memory layout.